### PR TITLE
Redirect DllImport of hostpolicy to the main executable when it's embedded

### LINKED ISSUE #39907: Single-File host does not correctly redirect hostpolicy PInvokes
There's a managed component `AssemblyDependencyResolver` which is implemented by PInvoking into `hostpolicy`. Since in the new single-file the `hostpolicy` doesn't exist as a separate module, we need to redirect the PInvokes to the main module, just like we do for other statically linked libraries.

Note that this needs to work on Windows as well - which is different from the case we already implement which is Linux only.

Repro:
```C#
using System.Runtime.Loader;

namespace app
{
    class Program
    {
        static void Main(string[] args)
        {
            AssemblyDependencyResolver resolver = new AssemblyDependencyResolver("path.dll");
        }
    }
}
```

Run as a single-file self-contained console app. i.e. `dotnet publish -r win-x64 /p:PublishSingleFile=true`. The result is:
```
Unhandled exception. System.InvalidOperationException: Cannot load hostpolicy library. AssemblyDependencyResolver is currently only supported if the runtime is hosted through hostpolicy library.
 ---> System.DllNotFoundException: Dll was not found.
   at Interop.HostPolicy.corehost_set_error_writer(IntPtr errorWriter)
   at System.Runtime.Loader.AssemblyDependencyResolver..ctor(String componentAssemblyPath)
```

It will work if published as FDD single-file i.e. `dotnet publish -r win-x64 --self-contained false /p:PublishSingleFile=true` - this will now fail with error about not finding `path.dll`.

I verified that it behaves the same on Linux with the latest RC from installers.

### PR DESCRIPTION
**Pull request by vitek-karas**
For `singlefilehost` the `hostpolicy` is statically linked into the executable, so we need to apply similar logic to other statically linked libraries by redirecting `DllImport` loads to the main executable module.

Because this happens also on Windows, the coreclr itself doesn't know statically (unlike on Linux where we link coreclr statically as well, so we can hardcode this knowledge). So the hostpolicy must pass this information to the runtime via a new runtime property.

Tested both on Windows and Linux.

Fixes #39907

### CODE REVIEW

File: src/coreclr/src/vm/dllimport.cpp [hunk 1]
```diff
@@ -6328,6 +6331,19 @@ namespace
         }
 #endif
 
+        if (g_hostpolicy_embedded)
+        {
+#if defined(TARGET_LINUX)
```
Comment (am11) [inline src/coreclr/src/vm/dllimport.cpp:6336]: Would it be a good idea to invert this condition to make this logic future proof; such that all OS except Windows get `libhostpolicy`?

Comment (VSadov) [reply src/coreclr/src/vm/dllimport.cpp:6336]: Yes, I think everything but Windows should use `libhostpolicy`

Comment (VSadov) [reply src/coreclr/src/vm/dllimport.cpp:6336]: Is this actually the name used in this context? - not `libhostpolicy.so` or something with a path?

In the case above we could, in theory, see different file names, but we intentionally only recognize the form used in `Interop.Libraries.cs`.
Jut wonder whether there is anything to worry about hostpolicy naming.

Comment (vitek-karas) [reply src/coreclr/src/vm/dllimport.cpp]: The story is the same - we don't expect any other code but corelib to call into hostpolicy directly, so these strings need to match what's in `Interop.Libraries`, but nothing else.

I fixed the condition to be for Windows - should also fix the build failures in CI

**Review by VSadov — APPROVED** [review]
LGTM

File: src/coreclr/src/vm/dllimport.cpp [hunk 2]
```diff
@@ -6328,6 +6331,19 @@ namespace
         }
 #endif
 
+        if (g_hostpolicy_embedded)
+        {
+#ifdef TARGET_WINDOWS
+            if (wcscmp(wszLibName, W("hostpolicy.dll")) == 0) {
```
Comment (am11) [inline src/coreclr/src/vm/dllimport.cpp:6337]: nit: `{` on its own line for consistency.

**Comment by vitek-karas** [comment]
@BruceForstall can you please take a look why the formatting job is failing on this PR? The changes don't touch JIT code. And the generated patch file is empty (0 bytes).

**Comment by am11** [comment]
Windows formatting job failure is unrelated to PR changes, failing in every PR. See https://github.com/dotnet/runtime/issues/40034.

**Comment by vitek-karas** [comment]
Thanks @am11  - I'm trying to figure out the other failures now...

**Comment by vitek-karas** [comment]
Something weird is happening. One of the tests is failing with:
```
Launching 'C:\h\w\B8170A18\w\AC3D09A7\e\Interop\COM\NativeClients\Primitives\COMClientPrimitives.exe'...

Expected: 100
Actual: 3
END EXECUTION - FAILED
FAILED
```

But looking at the code https://github.com/dotnet/runtime/blob/7242e14c534439d8b3aac9f90ce04224e1bdda70/src/tests/Interop/COM/NativeClients/Primitives/Client.cpp#L24-L45

There's no way it returns exit code 3.

**Comment by am11** [comment]
The `Expected: 3` is indeed very fishy. Going by this investigation: https://github.com/dotnet/runtime/pull/33733#issuecomment-610565638, it might be a similar intermittent assert is failing in runtime, GUI popping up on the build agent causing exit code 3. Maybe rerunning the CI will fix this error.
`git commit -m "Re-run CI" --alow-empty && git push`

**Comment by vitek-karas** [comment]
It is an assert - but seems valid - debugging through it right now.

**Comment by vitek-karas** [comment]
The test is mocking hostpolicy - so maybe there's something going on there (although I would not expect to affect non-single-file builds at all, given that the embedded variables should be false in that case)

**Comment by vitek-karas** [comment]
Found it - uninitialized variable (I've been working in C# for way too long now, forgot all of the C++ gotchas)... I can only wish for default compiler settings which would definitely show this as a warning.

**Comment by am11** [comment]
Heh, it's C++ which first says `Values of type bool are either true or false.` then later `... uninitialized automatic variable, might cause it to behave as if it is neither true nor false.` 🙈

**Comment by vitek-karas** [comment]
In this case (on my machine anyway) it initialized to magical 204 (literally) - and since conditions treat anything non-zero as true....
Anyway - WARNINGS - but I guess that would be a project for like 3 months to clean up coreclr to be able to turn on at least some C++ warnings.

**Comment by vitek-karas** [comment]
Networking test failures are being fixed in https://github.com/dotnet/runtime/pull/40070.

**Comment by jkotas** [comment]
> WARNINGS - but I guess that would be a project for like 3 months to clean up coreclr to be able to turn on at least some C++ warnings.

We have all reasonable C++ warnings turned on. C++ compilers do not seem to provide warning to catch this type of problem. Neither `MSVC` w/ `/W4` nor `clang` w/ `-Wall` warn for this code:
```
extern void g(bool* p);

void f()
{
   bool f;
   g(&f);
}
```

**Comment by vitek-karas** [comment]
It's great that we have warnings on... and not so great that the compiler doesn't have a way to warn on this (I guess it is a hard problem in C++). I definitely didn't mean to downplay our efforts on the codebase...

**Pull request merged**